Accept tokenId's for spending and swapping

swansontec opened the pull request:
We will upgrade the `currencyCode` into the `tokenId` if not provided, and vice-versa so the plugin remains compatible as well.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203287918505521

swansontec commented on src/core/account/custom-tokens.js:
```diff
@@ -247,3 +247,27 @@ export async function saveCustomTokens(
+
+  const tokenId =
+    info.tokenId ??
+    Object.keys(allTokens).find(
```
Oops. This doesn’t handle when the passed currency code is for the chain itself, like “BTC”

paullinator commented on src/types/error.js:
```diff
@@ -264,19 +264,45 @@ export class SwapCurrencyError extends Error {
+    }
+
     super(
       `${swapInfo.displayName} does not support ${fromCurrency} to ${toCurrency}`
```
Should probably show the tokenIds as well to differentiate what was trying to be used. That's the purpose of the error.

swansontec replied:
The error is meant to be translated by the GUI into something more usable, which is why we put the tokenId's into the data part. I can adjust the message to include those as well, in case the GUI doesn't do the translation.

paullinator replied:
That would be great. Also helps for the logs that gets spewed.

paullinator commented on src/types/error.js:
```diff
@@ -264,19 +264,45 @@ export class SwapCurrencyError extends Error {
+          ? request.fromWallet.currencyConfig.allTokens[request.fromTokenId]
+              .currencyCode
+          : request.fromCurrencyCode ??
+            request.fromWallet.currencyInfo.currencyCode
```
why fallback to `request.fromWallet.currencyInfo.currencyCode`? Shouldn't the request.fromCurrencyCode always exist?

swansontec replied:
The type system demands it. But I could put `''` or something if that would be simpler.

paullinator replied:
I think some invalid value would be better so we don't imply that is an expected code-flow. `` would be fine.

swansontec commented:
@paullinator I have had to do some refactoring to address your issues. I moved the `upgradeCurrencyCode` into its own file, then took advantage of it inside the `SwapCurrencyError` constructor. This new file needs to be included in the Flow -> TS conversion script, since errors.js depends on it. Finally, I fixed `upgradeCurrencyCode` to address cases where the currency code matches the parent currency.

paullinator commented:
looks good.

swansontec commented:
/autosquash
